### PR TITLE
Prevent pops from sinking in SimplifyLocals

### PR DESCRIPTION
**Pull request by aheejin**
This prevents `exnref.pop`s from being sinked and separated from
`catch`. For example,
```wast
(try
  (do)
  (catch
    (local.set $0 (exnref.pop))
    (call $foo
      (i32.const 3)
      (local.get $0)
    )
  )
)
```
Here, if we sink `exnref.pop` to remove `local.set $0` and
`local.get $0`, it becomes this:
```wast
(try
  (do)
  (catch
    (nop)
    (call $foo
      (i32.const 3)
      (exnref.pop)
    )
  )
)
```
This move was possible because `i32.const 3` does not have any side
effects. But this is incorrect because now `exnref.pop` does not follow
right after `catch`.

To prevent this, this patch checks this case in `canSink` in
SimplifyLocals. When we encountered a similar case in CodeFolding, we
prevented every expression that contains `Pop` anywhere in it from being
moved, which was too conservative. This adds `danglingPop` property in
`EffectAnalyzer`, so that only pops that are not enclosed within a
`catch` count as 'dangling pops` and we only prevent those pops from
being moved or sinked.

### CODE REVIEW

File: src/ir/effects.h [hunk 1]
```diff
@@ -470,7 +484,12 @@ struct EffectAnalyzer
   }
   void visitNop(Nop* curr) {}
   void visitUnreachable(Unreachable* curr) { branchesOut = true; }
-  void visitPop(Pop* curr) { calls = true; }
+  void visitPop(Pop* curr) {
+    // When EH is enabled, any call can throw.
```
Comment (kripken) [inline src/ir/effects.h:488]: `call` seems irrelevant here?

File: src/ir/effects.h [hunk 2]
```diff
@@ -470,7 +484,12 @@ struct EffectAnalyzer
...
+    if (features.hasExceptionHandling() && catchDepth == 0) {
```
Comment (kripken) [inline src/ir/effects.h:489]: why check the feature here - if a pop exists, doesn't that imply we must have exceptions enabled?

File: src/passes/SimplifyLocals.cpp [hunk 3]
```diff
@@ -411,6 +411,13 @@ struct SimplifyLocals
     if (set->isTee()) {
       return false;
     }
+    // We cannot move expression containing exnref.pops that are not enclosed in
```
Comment (kripken) [inline src/passes/SimplifyLocals.cpp:414]: `expression => expressions`

File: src/passes/SimplifyLocals.cpp [hunk 4]
```diff
@@ -411,6 +411,13 @@ struct SimplifyLocals
...
+    // 'catch', because 'exnref.pop' should follow right after 'catch'.
+    if (EffectAnalyzer(
+          this->getPassOptions(), this->getModule()->features, set->value)
+          .danglingPop) {
```
Comment (kripken) [inline src/passes/SimplifyLocals.cpp:418]: I think we call `invalidates()` which would do what we need automatically. that is, it's ok to sink if we don't cross over another expression. and so this check is not needed.

File: src/ir/effects.h (hunk 1 above)
Comment (aheejin) [reply src/ir/effects.h:488]: Oh, thanks, deleted the comment. I was copy-pasting and didn't end up deleting that I guess. Or are you talking about that I deleted `calls = true;`? I'm not sure why `calls` was set to true for `pop`s in the first place, it was that way when `pop` support was first added by you. I guess you did that to prevent `pop`s from being moved willy-nilly when they should't be moved I guess. But now we have `danglingPop`, which sets `hasSideEffects()` to true, I think the same effect can be achieved by that.

File: src/ir/effects.h (hunk 2 above)
Comment (aheejin) [reply src/ir/effects.h:489]: Hmm, yeah, right. Deleted it.

File: src/passes/SimplifyLocals.cpp (hunk 3 above)
Comment (aheejin) [reply src/passes/SimplifyLocals.cpp:414]: Done

File: src/passes/SimplifyLocals.cpp (hunk 4 above)
Comment (aheejin) [reply src/passes/SimplifyLocals.cpp:418]: No, it's not about invalidating side effects. What we crossed over in my test case `$pop-cannot-be-sinked` is `(i32.const 3)`, which should be OK to be reordered with anything without changing any side effects. But `pop` cannot be reordered even with `i32.const` because `pop` should follow right after `catch`. Deleting this code make that test fail.

Comment (aheejin) [reply src/passes/SimplifyLocals.cpp:418]: Oh, come to think of it, I guess I should change `invalidates()` method instead, so that it returns true when either side of expression is `pop`, regardless of what the other side is? Is this what you meant? I'll do that. Please let me know if this is not what you meant.

Comment (aheejin) [reply src/passes/SimplifyLocals.cpp:418]: Hmm, it's weird, we're already checking that in `invalidate`... I'll investigate.

Comment (aheejin) [reply src/passes/SimplifyLocals.cpp:418]: So the test code that's failing is here: (Here `local.set` should NOT be sinked)
```wast
(catch
  (local.set $0 (exnref.pop))
  (call $foo
    (i32.const 3)
    (local.get $0)
  )
)
```
So `invalidates` check is called [here](https://github.com/WebAssembly/binaryen/blob/beafd4721df6233556dfa330342ad10de7c0b9f5/src/passes/SimplifyLocals.cpp#L393-L395), and it is only called `checkPost` returns true. If that has to be called for our `(local.set (exnref.pop))`, at this point `original` should be `(i32.const 3)`. And `sinkables` [here](https://github.com/WebAssembly/binaryen/blob/beafd4721df6233556dfa330342ad10de7c0b9f5/src/passes/SimplifyLocals.cpp#L286) would contain `(local.set (exnref.pop))`. But [`checkPost`](https://github.com/WebAssembly/binaryen/blob/beafd4721df6233556dfa330342ad10de7c0b9f5/src/ir/effects.h#L234-L240) does not return true for `(i32.const 3)`, because it does not have any side effects or anything. I'm not very familiar with how SimplifyLocals works, so I might be misunderstanding something.

File: src/ir/effects.h (hunk 1 above)
Comment (kripken) [reply src/ir/effects.h]: Yes, I meant the comment looked wrong.

File: src/passes/SimplifyLocals.cpp (hunk 4 above)
Comment (kripken) [reply src/passes/SimplifyLocals.cpp:418]: Oh right, it doesn't just check invalidates, it incrementally calls `checkPre/Post`. So yes, how you did it is the only way.

File: src/passes/SimplifyLocals.cpp [hunk 5]
```diff
@@ -411,6 +411,13 @@ struct SimplifyLocals
     if (set->isTee()) {
       return false;
     }
+    // We cannot move expressions containing exnref.pops that are not enclosed
+    // in 'catch', because 'exnref.pop' should follow right after 'catch'.
+    if (EffectAnalyzer(
```
Comment (kripken) [inline src/passes/SimplifyLocals.cpp:416]: how about checking for the exceptions feature here, to avoid computing all side effects when it is disabled?

Comment (aheejin) [reply src/passes/SimplifyLocals.cpp:416]: Done
